src/vm/checker/check_db.rs

```diff
@@ -0,0 +1,113 @@
+
+        contract_db.check_schema();
+
+        contract_db
+    }
+
+    pub fn memory() -> AnalysisDatabase {
+        AnalysisDatabase::initialize(":memory:")
+    }
+
+    pub fn open(filename: &str) -> AnalysisDatabase {
+        let contract_db = AnalysisDatabase::inner_open(filename);
+
+        contract_db.check_schema();
+        contract_db
+    }
+
+    pub fn check_schema(&self) {
+        let sql = "SELECT sql FROM sqlite_master WHERE name=?";
+        let _: String = self.conn.query_row(sql, &["type_analysis_table"],
+                                            |row| row.get(0))
+            .expect("Bad schema in analysis db initialization.");
+    }
+
+    pub fn inner_open(filename: &str) -> AnalysisDatabase {
+        let conn = Connection::open(filename)
+            .expect("Failure to open analysis db.");
+        AnalysisDatabase {
+            conn: conn
+        }
+    }
+
+    pub fn execute<P>(&mut self, sql: &str, params: P) -> usize
+    where
+        P: IntoIterator,
+        P::Item: ToSql {
+        self.conn.execute(sql, params)
+            .expect(SQL_FAIL_MESSAGE)
+    }
+
+    fn query_row<T, P, F>(&self, sql: &str, params: P, f: F) -> Option<T>
+    where
+        P: IntoIterator,
+        P::Item: ToSql,
+        F: FnOnce(&Row) -> T {
+        self.conn.query_row(sql, params, f)
+            .optional()
+            .expect(SQL_FAIL_MESSAGE)
+    }
+
+    fn load_contract(&self, contract_name: &str) -> Option<ContractAnalysis> {
+        let result: Option<String> = self.query_row(
+            "SELECT analysis FROM type_analysis_table WHERE contract_name = ?",
+            &[contract_name],
+            |row| row.get(0));
+        match result {
+            Some(contract) => Some(ContractAnalysis::deserialize(&contract)),
+            None => None
+        }
+    }
+
+    pub fn get_public_function_type(&self, contract_name: &str, function_name: &str) -> CheckResult<FunctionType> {
+        let contract = self.load_contract(contract_name)
+            .ok_or(CheckError::new(CheckErrors::NoSuchContract(contract_name.to_string())))?;
+        Ok(
+            contract.get_public_function_type(function_name)
+                .ok_or(CheckError::new(CheckErrors::NoSuchPublicFunction(contract_name.to_string(),
+                                                                         function_name.to_string())))?
+                .clone())
+    }
+
+    pub fn get_map_type(&self, contract_name: &str, map_name: &str) -> CheckResult<(TypeSignature, TypeSignature)> {
+        let contract = self.load_contract(contract_name)
+            .ok_or(CheckError::new(CheckErrors::NoSuchContract(contract_name.to_string())))?;
+        let map_type = contract.get_map_type(map_name)
+            .ok_or(CheckError::new(CheckErrors::NoSuchMap(map_name.to_string())))?;
+        Ok(map_type.clone())
+    }
+
+    pub fn insert_contract(&mut self, contract_name: &str, contract: &ContractAnalysis) -> CheckResult<()> {
```

Comment:
How are contract names decided?  All I can see in this PR is that they're decided via the `main()` method.  Is this still TBD, per our earlier discussions on having human-readable names for contracts?  How will contract-instantiation transactions pass the contract name?

Reply:
Yeah, it's left as "TBD" still in this PR.